MoreLinq.Test/ScanByTest.cs

```diff
@@ -0,0 +1,120 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Leandro F. Vieira (leandromoh). All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class ScanByTest
+    {
+        [Test]
+        public void ScanByIsLazy()
+        {
+            new BreakingSequence<string>().ScanBy(
+                BreakingFunc.Of<string, int>(),
+                BreakingFunc.Of<int, char>(),
+                BreakingFunc.Of<char, int, string, char>());
+        }
+
+        [Test]
+        public void ScanBy()
+        {
+            var source = new[]
+            {
+                "ana",
+                "beatriz",
+                "carla",
+                "bob",
+                "davi",
+                "adriano",
+                "angelo",
+                "carlos"
+            };
+
+            var result =
+                from x in
+                    source.ScanBy(
+                        item => item.First(),
+                        key => (Element: default(string), Key: key, State: (int)key - 1),
+                        (state, key, item) => (Element: item, Key: char.ToUpper(key), State: state.State + 1))
```

Comment:
- Replace `ToUpper` with the more deterministic `ToUpperInvariant`.
- The tuple element names here are redundant. I think it reads equally clear if we remove them, as in `(item, char.ToUpper(key), state.State + 1)`.